### PR TITLE
Added annotation typing to models._utils

### PR DESCRIPTION
**Pull request by frgfm**
Hello folks :wave: 

As per #2025, annotation typing are welcome in torchvision. So I started with `torchvision.models._utils` in this PR. A few points of attention:
- I picked `Callable[..., nn.Module]` for module constructors (passing `BatchNorm2d` for instance), I'm unsure whether that is good choice? any recommendation?
-  I only used black styling in class constructors for now, the rest are only annotation typing additions. Happy to change everything to black linting if you prefer.

Any feedback is welcome!

### CODE REVIEW

File: torchvision/models/_utils.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 from collections import OrderedDict
 
 import torch
+from torch import Tensor
```
Comment (pmeier) [inline torchvision/models/_utils.py:4]: Given that we don't need to use this in so many places and we already importing `torch`, couldn't we simply use `torch.Tensor`?

File: torchvision/models/_utils.py [hunk 2]
```diff
@@ -57,7 +58,7 @@ def __init__(self, model, return_layers):
         super(IntermediateLayerGetter, self).__init__(layers)
         self.return_layers = orig_return_layers
 
-    def forward(self, x):
+    def forward(self, x: Tensor) -> Tensor:
```
Comment (pmeier) [inline torchvision/models/_utils.py:61]: Shouldn't that be `-> Dict[str, torch.Tensor]`?

File: torchvision/models/alexnet.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import torch
+from torch import Tensor
```
Comment (pmeier) [inline torchvision/models/alexnet.py:2]: Same as above.

File: torchvision/models/alexnet.py [hunk 4]
```diff
@@ -41,15 +42,15 @@ def __init__(self, num_classes=1000):
...
+    def forward(self, x: Tensor) -> Tensor:
         x = self.features(x)
         x = self.avgpool(x)
         x = torch.flatten(x, 1)
         x = self.classifier(x)
         return x
 
 
-def alexnet(pretrained=False, progress=True, **kwargs):
+def alexnet(pretrained: bool = False, progress: bool = True, **kwargs) -> AlexNet:
```
Comment (pmeier) [inline torchvision/models/alexnet.py:53]:
```suggestion
def alexnet(pretrained: bool = False, progress: bool = True, **kwargs: Any) -> AlexNet:
```

File: torchvision/models/densenet.py [hunk 5]
```diff
@@ -35,41 +43,36 @@ def __init__(self, num_input_features, growth_rate, bn_size, drop_rate, memory_e
...
-        # type: (List[Tensor]) -> Tensor
+    def bn_function(self, inputs: List[Tensor]) -> Tensor:
         concated_features = torch.cat(inputs, 1)
         bottleneck_output = self.conv1(self.relu1(self.norm1(concated_features)))  # noqa: T484
         return bottleneck_output
 
     # todo: rewrite when torchscript supports any
-    def any_requires_grad(self, input):
-        # type: (List[Tensor]) -> bool
+    def any_requires_grad(self, input: List[Tensor]) -> Tensor:
```
Comment (pmeier) [inline torchvision/models/densenet.py:52]:
```suggestion
    def any_requires_grad(self, input: List[Tensor]) -> bool:
```

File: torchvision/models/densenet.py [hunk 6]
```diff
@@ -35,41 +43,36 @@ def __init__(self, num_input_features, growth_rate, bn_size, drop_rate, memory_e
...
     @torch.jit._overload_method  # noqa: F811
-    def forward(self, input):
-        # type: (Tensor) -> (Tensor)
+    def forward(self, input: Tensor) -> Tensor:
         pass
 
     # torchscript does not yet support *args, so we overload method
     # allowing it to take either a List[Tensor] or single Tensor
-    def forward(self, input):  # noqa: F811
+    def forward(self, input: Union[Tensor, List[Tensor]]) -> Tensor:  # noqa: F811
```
Comment (pmeier) [inline torchvision/models/densenet.py:75]: `torch.script` is not able to handle `Union`. We need to annotate it as `torch.Tensor` for now and add an ignore for `mypy`.

File: torchvision/models/densenet.py [hunk 7]
```diff
@@ -215,15 +234,15 @@ def _load_state_dict(model, model_url, progress):
     model.load_state_dict(state_dict)
 
 
-def _densenet(arch, growth_rate, block_config, num_init_features, pretrained, progress,
-              **kwargs):
+def _densenet(arch: str, growth_rate: int, block_config: Tuple[int], num_init_features: int,
+              pretrained: bool, progress: bool, **kwargs) -> DenseNet:
```
Comment (pmeier) [inline torchvision/models/densenet.py:238]:
```suggestion
              pretrained: bool, progress: bool, **kwargs: Any) -> DenseNet:
```

File: torchvision/models/densenet.py [hunk 8]
```diff
@@ -215,15 +234,15 @@ def _load_state_dict(model, model_url, progress):
...
     model = DenseNet(growth_rate, block_config, num_init_features, **kwargs)
     if pretrained:
         _load_state_dict(model, model_urls[arch], progress)
     return model
 
 
-def densenet121(pretrained=False, progress=True, **kwargs):
+def densenet121(pretrained: bool = False, progress: bool = True, **kwargs) -> DenseNet:
```
Comment (pmeier) [inline torchvision/models/densenet.py:245]:
```suggestion
def densenet121(pretrained: bool = False, progress: bool = True, **kwargs: Any) -> DenseNet:
```

File: torchvision/models/densenet.py [hunk 9]
```diff
@@ -237,7 +256,7 @@ def densenet121(pretrained=False, progress=True, **kwargs):
                      **kwargs)
 
 
-def densenet161(pretrained=False, progress=True, **kwargs):
+def densenet161(pretrained: bool = False, progress: bool = True, **kwargs) -> DenseNet:
```
Comment (pmeier) [inline torchvision/models/densenet.py:259]:
```suggestion
def densenet161(pretrained: bool = False, progress: bool = True, **kwargs: Any) -> DenseNet:
```

File: torchvision/models/densenet.py [hunk 10]
```diff
@@ -251,7 +270,7 @@ def densenet161(pretrained=False, progress=True, **kwargs):
                      **kwargs)
 
 
-def densenet169(pretrained=False, progress=True, **kwargs):
+def densenet169(pretrained: bool = False, progress: bool = True, **kwargs) -> DenseNet:
```
Comment (pmeier) [inline torchvision/models/densenet.py:273]:
```suggestion
def densenet169(pretrained: bool = False, progress: bool = True, **kwargs: Any) -> DenseNet:
```

File: torchvision/models/densenet.py [hunk 11]
```diff
@@ -265,7 +284,7 @@ def densenet169(pretrained=False, progress=True, **kwargs):
                      **kwargs)
 
 
-def densenet201(pretrained=False, progress=True, **kwargs):
+def densenet201(pretrained: bool = False, progress: bool = True, **kwargs) -> DenseNet:
```
Comment (pmeier) [inline torchvision/models/densenet.py:287]:
```suggestion
def densenet201(pretrained: bool = False, progress: bool = True, **kwargs: Any) -> DenseNet:
```

**Review by pmeier — CHANGES_REQUESTED** [review]
Hey @frgfm, thanks a lot for the PR! Could you please split this PR into multiple as it is very hard to review with that many changes. I think it would be best to do a separate PR for every model / file. 

I've reviewed the first three files and found a couple of things.

**Comment by frgfm** [comment]
@pmeier sure, I updated this PR to focus on `models._utils`. You got a dozen PRs incoming then haha

**Comment by pmeier** [comment]
That is alright. I'll go through them one by one.

**Review by pmeier — CHANGES_REQUESTED** [review]
`mypy` CI [is failing](https://app.circleci.com/pipelines/github/pytorch/vision/4535/workflows/507b55a2-b19a-42b6-94c2-597db452e110/jobs/272328) with
```
torchvision/models/_utils.py:60: error:
Signature of "forward" incompatible with supertype "ModuleDict" [override]
        def forward(self, x: torch.Tensor) ->...
```

**Comment by pmeier** [comment]
@frgfm Removing the annotation is not the way to go here. Either fix it or if it is unfixable (which might be the case here) use the correct annotations and add a `#type: ignore[override]` to the line.

**Comment by pmeier** [comment]
The test failures are related. I'll look into it and get back to you.

**Comment by pmeier** [comment]
@frgfm It seems torchscript is the offender here. I suspect it can't handle the `type: ignore` comment. Could you remove that to see if my suspicion is right?

If it is I'm not sure to move forward. We can't make torchscript *and* `mypy` happy.

**Comment by frgfm** [comment]
@pmeier Yes, I figured out the same thing yesterday unfortunately.
When I remove the ignore, mypy throws the following:
```
error: Signature of "forward" incompatible
with supertype "ModuleDict"  [override]
```

Let me know which way you prefer to proceed with!

**Comment by pmeier** [comment]
In general torchscript > mypy holds. Thus, while the type hints are nice to have, passing tests for torchscript is a requirement. Could you simply remove the `type: ignore` comment and push that change? It is not important for now if `mypy` fails, but the unit tests need to pass.

**Comment by fmassa** [comment]
Yes.

The situation with torchscript is delicate, and if we need to chose between mypy and torchscript, torchscript wins.

In general, all torchvision models already have type annotation in the forward because they all support torchscript. The difference with mypy is that torchscript assumes that
- if no input type is passed to an argument, it's assumed to be a tensor
- return types are inferred

So if you want to play safe, I would say to only add type annotations to the constructor of the models, or the model builder functions. This way it won't affect torchscript, but mypy won't probably be happy either.

Which means that it might be hard to enable full mypy testing to the models.

**Comment by pmeier** [comment]
> This way it won't affect torchscript, but mypy won't probably be happy either.

Right now, we allow untyped definitions. Thus, if we simply remove the annotations from `forward`, `mypy` should be happy.

**Review by pmeier — APPROVED** [review]
LGTM. Thanks a lot @frgfm. Note that the further process will be delayed, since we are currently preparing the next release of `torchvision`.

**Comment by frgfm** [comment]
> LGTM. Thanks a lot @frgfm. Note that the further process will be delayed, since we are currently preparing the next release of `torchvision`.

Should I update 
```
from torch.jit.annotations import Dict
```
to 
```
from typing import Dict
```
though?

**Comment by pmeier** [comment]
Yes that would be nice!

**Review by fmassa — APPROVED** [review]
Thanks!
